kubetest2-gke/deployer/network.go

```diff
@@ -38,41 +39,115 @@ etag: %s
 	}
 
 	numProjects := len(d.Projects)
 	if numProjects == 0 {
 		numProjects = d.BoskosProjectsRequested
 	}
-	// For single project, no other verification is needed.
-	if numProjects == 1 {
-		return nil
-	}
 
-	if d.Network == "default" {
-		return errors.New("the default network cannot be used for multi-project profile")
+	// Verify for multi-project profile.
+	if numProjects > 1 {
+		if d.Network == "default" {
+			return errors.New("the default network cannot be used for multi-project profile")
+		}
+
+		if len(d.SubnetworkRanges) != (numProjects-1)*d.totalTryCount {
+			return fmt.Errorf("the number of subnetwork ranges provided "+
+				"should be the same as the number of service projects times the total try count : %d!=%d", len(d.SubnetworkRanges), (numProjects-1)*d.totalTryCount)
+		}
+
+		if err := d.validateSubnetRanges(); err != nil {
+			return err
+		}
 	}
 
-	if len(d.SubnetworkRanges) != numProjects-1 {
-		return fmt.Errorf("the number of subnetwork ranges provided "+
-			"should be the same as the number of service projects: %d!=%d", len(d.SubnetworkRanges), numProjects-1)
+	if err := d.internalizeNetworkFlags(numProjects); err != nil {
+		return err
 	}
 
-	for _, sr := range d.SubnetworkRanges {
-		parts := strings.Split(sr, " ")
-		if len(parts) != 3 {
+	return nil
+}
+
+func (d *Deployer) validateSubnetRanges() error {
+	// The subnets are passed in a list, each containing groups of 3 CIDR ranges.
+	// We need to verify there are no overlaps within the entire group.
+	var allSubnetRanges []string
+	for _, subnet := range d.SubnetworkRanges {
+		ranges := strings.Split(subnet, " ")
+		if len(ranges) != 3 {
 			return fmt.Errorf("the provided subnetwork range %s is not in the right format, should be like "+
-				"10.0.4.0/22 10.0.32.0/20 10.4.0.0/14", sr)
+				"10.0.4.0/22 10.0.32.0/20 10.4.0.0/14", subnet)
+		}
+
+		allSubnetRanges = append(allSubnetRanges, ranges...)
+	}
+
+	if err := assertNoOverlaps(allSubnetRanges); err != nil {
+		return fmt.Errorf("error in subnetwork ranges: %v", err)
+	}
+
+	return nil
+}
+
+func assertNoOverlaps(ranges []string) error {
+	var allRanges []*net.IPNet
+	for _, rangeString := range ranges {
+		_, thisRange, err := net.ParseCIDR(rangeString)
+		if err != nil {
+			return fmt.Errorf("error parsing %q into a CIDR: %w", rangeString, err)
+		}
+		for _, previousRange := range allRanges {
+			if areOverlapping(thisRange, previousRange) {
+				return fmt.Errorf("overlap found within the provided ranges: (%v, %v)", thisRange, previousRange)
+			}
 		}
+		allRanges = append(allRanges, thisRange)
 	}
+	return nil
+}
 
+func areOverlapping(r1, r2 *net.IPNet) bool {
+	return r1.Contains(r2.IP) || r2.Contains(r1.IP)
+}
+
+func (d *Deployer) internalizeNetworkFlags(numProjects int) error {
+	d.subnetworkRangesInternal = make([][]string, d.totalTryCount)
+	for tc := 0; tc < d.totalTryCount; tc++ {
+		d.subnetworkRangesInternal[tc] = make([]string, numProjects-1)
+		for p := 0; p < numProjects-1; p++ {
```

Comment:
This n-1 indexing for hostProject v/s service projects is starting to become magical.

I would suggest breaking down the logic into a `SubNetwork` struct which would store the project info in a follow up.